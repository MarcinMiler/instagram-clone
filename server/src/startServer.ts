import 'reflect-metadata'
import { importSchema } from 'graphql-import'
import { GraphQLServer } from 'graphql-yoga'
import * as path from 'path'
import * as fs from 'fs'
import * as jwt from 'jsonwebtoken'
import { mergeSchemas, makeExecutableSchema } from 'graphql-tools'
import { GraphQLSchema } from 'graphql'
import { createTypeormConn } from './utils/createTypeormConn'
import SECRET from './utils/SECRET'

const Expo = require('expo-server-sdk')

const options = {
    port: process.env.NODE_ENV === 'test' ? 0 : 4000,
    endpoint: '/graphql',
    cors: {
        origin: '*'
    }
}

const addUser = (req: any, _: any, next: any) => {
    const token = req.headers.token
    if (token) {
        const user = jwt.verify(token, SECRET)
        req.user = user
    }
    next()
}

export const startServer = async () => {
    const schemas: GraphQLSchema[] = []
    const folders = fs.readdirSync(path.join(__dirname, './modules'))
    folders.forEach(folder => {
        const { resolvers } = require(`./modules/${folder}/resolvers`)
        const typeDefs = importSchema(
            path.join(__dirname, `./modules/${folder}/schema.graphql`)
        )
        schemas.push(makeExecutableSchema({ resolvers, typeDefs }))
    })

    const expo = new Expo()

    const server = new GraphQLServer({
        schema: mergeSchemas({ schemas }),
        context: (req: any) => {
            if (req.request.user) {
                return {
                    user: req.request.user.id,
                    expo
                }
            }
            return
        }
    })
    server.use(addUser)

    await createTypeormConn()

    const app = await server.start(options)

    return app
}
