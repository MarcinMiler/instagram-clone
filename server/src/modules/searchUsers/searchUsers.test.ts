import { request, GraphQLClient } from 'graphql-request'
import { startServer } from '../../startServer'
import {
    loginMutationWithToken,
    registerMutation,
    searchUsersQuery
} from '../../utils/mutations'

let getHost = () => ''
let token = ''

const email = 'm@m.com'
const password = 'mm'
const fullname = 'Marcin Miler'
const username = 'Marcinek'

beforeAll(async () => {
    const app = await startServer()
    const { port } = app.address()
    getHost = () => `http://127.0.0.1:${port}/graphql`

    await request(
        getHost(),
        registerMutation(email, password, fullname, username)
    )

    await request(
        getHost(),
        registerMutation('a@a.com', 'aa', 'Angelika Miler', 'Angelaaa')
    )

    const login: any = await request(
        getHost(),
        loginMutationWithToken(email, password)
    )
    token = login.login.token
})

describe('Query search Users', async () => {
    it('should all users', async () => {
        const client = new GraphQLClient(getHost(), {
            headers: {
                token
            }
        })

        const response: any = await client.request(searchUsersQuery(''))

        expect(response.searchUsers).toHaveLength(2)
        expect(response).toMatchSnapshot()
    })

    it('should return user by username pattern', async () => {
        const client = new GraphQLClient(getHost(), {
            headers: {
                token
            }
        })

        const response: any = await client.request(searchUsersQuery('Marcinek'))

        expect(response.searchUsers).toHaveLength(1)
        expect(response).toMatchSnapshot()
    })
})
