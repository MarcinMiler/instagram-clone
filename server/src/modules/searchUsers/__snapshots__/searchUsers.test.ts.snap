// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query search Users should all users 1`] = `
Object {
  "searchUsers": Array [
    Object {
      "email": "m@m.com",
      "fullname": "Marcin Miler",
      "id": "1",
      "photos": null,
      "username": "Marcinek",
    },
    Object {
      "email": "a@a.com",
      "fullname": "Angelika Miler",
      "id": "2",
      "photos": null,
      "username": "Angelaaa",
    },
  ],
}
`;

exports[`Query search Users should return user by username pattern 1`] = `
Object {
  "searchUsers": Array [
    Object {
      "email": "m@m.com",
      "fullname": "Marcin Miler",
      "id": "1",
      "photos": null,
      "username": "Marcinek",
    },
  ],
}
`;
